model/map/src/main/java/org/keycloak/models/map/userSession/MapUserSessionProvider.java

```diff
@@ -0,0 +1,668 @@
+            userSessionEntityStream = getOfflineUserSessionEntityStream(realm, id)
+                    .map(userEntityToAdapterFunc(realm));
+        } else {
+            UserSessionModel userSession = getUserSession(realm, id);
+            userSessionEntityStream = userSession != null ? Stream.of(userSession) : Stream.empty();
+        }
+
+        return userSessionEntityStream
+                .filter(predicate)
+                .findFirst()
+                .orElse(null);
+    }
+
+    @Override
+    public long getActiveUserSessions(RealmModel realm, ClientModel client) {
+        ModelCriteriaBuilder<UserSessionModel> mcb = realmAndOfflineCriteriaBuilder(realm, false)
+                .compare(UserSessionModel.SearchableFields.CLIENT_ID, ModelCriteriaBuilder.Operator.EQ, client.getId());
+
+        LOG.tracef("getActiveUserSessions(%s, %s)%s", realm, client, getShortStackTrace());
+
+        return userSessionTx.getCount(mcb);
+    }
+
+    @Override
+    public Map<String, Long> getActiveClientSessionStats(RealmModel realm, boolean offline) {
+        ModelCriteriaBuilder<UserSessionModel> mcb = realmAndOfflineCriteriaBuilder(realm, offline);
+
+        LOG.tracef("getActiveClientSessionStats(%s, %s)%s", realm, offline, getShortStackTrace());
+
+        return userSessionTx.getUpdatedNotRemoved(mcb)
+                .map(userEntityToAdapterFunc(realm))
+                .map(UserSessionModel::getAuthenticatedClientSessions)
+                .map(Map::keySet)
+                .flatMap(Collection::stream)
+                .collect(Collectors.groupingBy(Function.identity(), Collectors.counting()));
+    }
+
+    @Override
+    public void removeUserSession(RealmModel realm, UserSessionModel session) {
+        Objects.requireNonNull(session, "The provided user session can't be null!");
+
+        ModelCriteriaBuilder<UserSessionModel> mcb = realmAndOfflineCriteriaBuilder(realm, false)
+                .compare(UserSessionModel.SearchableFields.ID, ModelCriteriaBuilder.Operator.EQ, UUID.fromString(session.getId()));
+
+        LOG.tracef("removeUserSession(%s, %s)%s", realm, session, getShortStackTrace());
+
+        userSessionTx.delete(UUID.randomUUID(), mcb);
+    }
+
+    @Override
+    public void removeUserSessions(RealmModel realm, UserModel user) {
+        ModelCriteriaBuilder<UserSessionModel> mcb = userSessionStore.createCriteriaBuilder()
+                .compare(UserSessionModel.SearchableFields.REALM_ID, ModelCriteriaBuilder.Operator.EQ, realm.getId())
+                .compare(UserSessionModel.SearchableFields.USER_ID, ModelCriteriaBuilder.Operator.EQ, user.getId());
+
+        LOG.tracef("removeUserSessions(%s, %s)%s", realm, user, getShortStackTrace());
+
+        userSessionTx.delete(UUID.randomUUID(), mcb);
+    }
+
+    @Override
+    public void removeAllExpired() {
+        LOG.tracef("removeAllExpired()%s", getShortStackTrace());
+    }
+
+    @Override
+    public void removeExpired(RealmModel realm) {
```

Comment:
Yes that's correct. When a session is never demanded again it will probably stay in the storage. I'm not sure if there are some deployments which rely on `removeExpired` method to do some kind of periodic cleanup of expired sessions.

Reply:
Should we solve this somehow? Or we are ok with that.

Reply:
I added this to the Jira https://issues.redhat.com/browse/KEYCLOAK-17884. We can discuss there.